### ISSUE TITLE
Value types and distance incompatibles with pHash

### ISSUE THREAD

**Issue by G10h4ck** [starkdg/mvptree#1]
Current MVPTree implementaion seems implemented in a way that it 
requires a bunch of modifications to be used together with pHash, in 
particular it uses double as value type and plain difference as distance, which 
seems totally incompatible with pHash hashes and their hamming distance

Using a bit of C++ templates I think I could generalize MVPTree in not much 
time but I might not have time to maintain a fork by myself in the future, if 
I do prepare a a pull request for generalized MVPTree would you review it and 
merge it, or the incompatible typing has been chosen on purpose?

**Comment by starkdg** [comment]
You are correct.  For use with phash types,  the member variables in DataType need to be set for each.  
This version is really just for experiments and such, since there is no way to save the tree to disk.  

sure, if you want to add c++ templates to make the datatype generic, I'd be glad to merge it in.  Just make
a new branch and I'll add it in as a  branch.  

For real applications, I started developing Redis modules.  There's a Redis-ImageScout repository that introduces
a mvptree as a redis native datatype with an 64-bit integer for data for a dct phash of the data.  
That's more useful, since you can scale up and save the entire tree on disk.
